sample/samplers.go

```diff
@@ -25,6 +25,10 @@ type Sampler struct {
 	grammar     *GrammarSampler
 }
 
+func (s *Sampler) SetGrammar(grammar *GrammarSampler) {
+	s.grammar = grammar
```

Comment:
I don't think this is thread safe. This handler is running asynchronously from the processing thread. This means that we could be sampling at the same time that the grammar is changed. I think there are two potential consequences:
- The data structure update is not atomic - you might see the pointer change before the data that it is pointing to has been moved. In practice there are probably enough memory barriers from the other code but it's not robust to rely on those side effects.
- There might be some tokens that are processed before the grammar change is seen, so they don't follow the the required constraints.

Reply:
I think a lock like that would help with the first issue but not really the second one.

The token generation code will continue running asynchronously until it hits a stopping point and just queue up the outputs in a channel, up to 100 tokens buffered currently. That means that depending on the relative speed of the threads, GPU, etc. that generation could be arbitrarily far ahead by the time the grammar gets set.